Add batch_partial implementation (#2585)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

  The code that checks for qubit wise commuting (QWC) got a performance boost that is noticable
  when many commuting paulis of the same type are measured.

* Added new transform `qml.batch_partial` which behaves similarly to `functools.partial` but supports batching in the unevaluated parameters.
  [(#2585)](https://github.com/PennyLaneAI/pennylane/pull/2585)

>   This is useful for batching circuit executions with some identical parameters but not others:

Reviewer: ```suggestion
  This is useful for executing a circuit with a batch dimension in some of its parameters:
```

  ```python
  dev = qml.device("default.qubit", wires=1)

  @qml.qnode(dev)
  def circuit(x, y):
     qml.RX(x, wires=0)
     qml.RY(y, wires=0)
     return qml.expval(qml.PauliZ(wires=0))
  ```
  ```pycon
  >>> batched_partial_circuit = qml.batch_partial(circuit, x=np.array(np.pi / 2))
  >>> y = np.array([0.2, 0.3, 0.4])
  >>> batched_partial_circuit(y=y)

>   tensor([0.69301172, 0.67552491, 0.65128847], requires_grad=True)

Reviewer: :heart_eyes:
